### PR TITLE
Add constructor for Filter

### PR DESCRIPTION
**Pull request by arberiii**
Fixes [issue #869](https://github.com/short-d/short/issues/869).

### CODE REVIEW

File: backend/app/usecase/search/search.go [hunk 1]
```diff
@@ -31,6 +31,9 @@ func (s Search) Search(query Query, filter Filter) (Result, error) {
 	resultCh := make(chan Result)
 	defer close(resultCh)
 
+	if len(filter.Resources) > len(filter.Orders) {
+		filter.Orders = appendDefaultOrders(filter.Orders, len(filter.Resources)-len(filter.Orders))
```
Comment (rohithbalaji123) [inline backend/app/usecase/search/search.go:35]: @arberiii , just curious how does this maintain the one to one correspondence of the resource and order? Should we think about extracting resource and order into a separate structure in [Filter](https://github.com/short-d/short/blob/master/backend/app/usecase/search/filter.go).

Comment (arberiii) [reply backend/app/usecase/search/search.go]: I agree that your suggestion seems better. One mismatch as you mention offline is if the package's client supplies 3 resources and 2 order by's for first and third resource.

File: backend/app/usecase/search/filter.go [hunk 2]
```diff
@@ -10,9 +10,14 @@ const (
 	User
 )
 
+// OrderedResource represents a type of searchable objects together with the order.
+type OrderedResource struct {
+	Resource Resource
+	Order    order.By
+}
```
Comment (byliuyang) [inline backend/app/usecase/search/filter.go:17]: I like the idea of combining them into a single struct. However, it does look weird especially when we are adding 1 more attribute in the future. Can we add `func NewFilter() (Filter, error)` and `(f *F) IsValid() bool` and return error in search function instead?

Comment (arberiii) [reply backend/app/usecase/search/filter.go:17]: Okay, let me implement your suggestion then we can compare all three alternatives

File: backend/app/usecase/search/filter.go [hunk 3]
```diff
@@ -16,3 +16,17 @@ type Filter struct {
...
 	Orders     []order.By
 }
+
+// IsValid checks if the resources and orders have one-to-one relation.
+func (f *Filter) IsValid() bool {
+	return len(f.Resources) == len(f.Orders)
+}
+
+// NewFilter creates Filter.
+func NewFilter(maxResults int, resources []Resource, orders []order.By) Filter {
```
Comment (byliuyang) [inline backend/app/usecase/search/filter.go:26]: We should throw error here is there is mismatch between resource and order

File: backend/app/usecase/search/search.go [hunk 4]
```diff
@@ -28,6 +28,10 @@ type Result struct {
 
 // Search finds resources based on specified criteria.
 func (s Search) Search(query Query, filter Filter) (Result, error) {
+	if !filter.IsValid() {
+		return Result{}, errors.New("filter is not valid")
```
Comment (byliuyang) [inline backend/app/usecase/search/search.go:32]:
```suggestion
		return Result{}, errors.New("invalid filter")
```

File: backend/app/usecase/search/filter_test.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
+package search
+
+import (
+	"testing"
+
+	"github.com/short-d/app/fw/assert"
+	"github.com/short-d/short/backend/app/usecase/search/order"
+)
+
+func TestCreatedTime_ArrangeShortLinks(t *testing.T) {
```
Comment (byliuyang) [inline backend/app/usecase/search/filter_test.go:10]:
```suggestion
func TestNewFilter(t *testing.T) {
```

Comment (arberiii) [reply backend/app/usecase/search/filter_test.go]: used it as template 😓 . Thanks for the catch.

**Pull request merged**